### PR TITLE
Proof-of-concept for testing shell sessions

### PR DESCRIPTION
**Pull request by zupo**
In https://github.com/nix-dot-dev/nix.dev/pull/248 I introduced a way to export code samples into files, and then test them.

This commit adds an additional improvement on top of that: to test shell session examples, command by command, and verify the output of commands matches the output given in the shell session example.

### CODE REVIEW

**Comment by zupo** [comment]
Don't merge yet. Apparently `find` does not fail if one of the `exec` commands fail. So we have test failures, but CI thinks everything is OK. Need to rewrite it. Probably to `xargs`.

File: run_code_block_tests.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
 set -euo pipefail
 
-find extracted -name "*.shell_session" -exec echo "running {}" \; -exec ./convert_shell_session_to_pexpect_test.py {} \;
-find extracted -name "test_*"  -exec echo "running {}" \; -execdir {} \;
+find extracted -name "*.shell_session" | xargs -I{} sh -c 'echo "running {}"; ./convert_shell_session_to_pexpect_test.py {};'
+find extracted -name "test_*" | xargs -I{} sh -c 'echo "running {}"; cd `dirname {}`; ./`basename {}`'
```
Comment (zupo) [inline run_code_block_tests.sh:4]: Hey @cheriimoya, is there a better way to do this?

**Comment by zupo** [comment]
> Don't merge yet. Apparently `find` does not fail if one of the `exec` commands fail. So we have test failures, but CI thinks everything is OK. Need to rewrite it. Probably to `xargs`.

This is now taken care of, CI fails on test failures. 

What's next is rewriting the test runner to not rely on my `darwin-shell.nix`.

File: run_code_block_tests.sh (hunk 1 above)
Comment (cheriimoya) [reply run_code_block_tests.sh:4]: this looks reasonable to me. the only thing you could change is to use finds `-print0` and xargs `-0` expression like this: `find -name "blah" -print0 | xargs -0 echo do stuff`

for an explanation, check out the second point on this list https://www.theunixschool.com/2012/11/examples-find-command-xargs-linux.html , pretty much to allow spaces in filenames etc

**Comment by domenkozar** [comment]
What's missing here to get this merged?

**Comment by zupo** [comment]
Fixing the CI.

**Comment by zupo** [comment]
Rebased off of latest master.

**Comment by domenkozar** [comment]
What about https://zimbatm.github.io/mdsh/ :laughing:

**Comment by domenkozar** [comment]
We now have https://github.com/OceanSprint/tesh

**Comment by zupo** [comment]
> We now have https://github.com/OceanSprint/tesh

Yep, I'll rewrite this PR.

**Comment by infinisil** [comment]
Are you still interested in working on this?

**Comment by infinisil** [comment]
Yeah long time no see, hope you're doing good!

I don't have any time to work on this myself either, but we were triaging issues in the documentation team meeting. Since you're planning to work on this, let's leave it open.

Note that this is very related to @mightyiam's recent work on a tester for `nix-repl` in https://github.com/NixOS/nix.dev/pull/672. Might be good to synchronize on that before diving in. I can recommend checking in with the documentation team [on Matrix](https://matrix.to/#/#docs:nixos.org) or [the meetings](https://github.com/NixOS/nix.dev/tree/master/maintainers#team-meetings) for that :)

**Comment by infinisil** [comment]
I'd say let's close this for now, can always be re-opened (or a new one opened) if work is being done.

**Pull request closed without merging**